packages/flutter_tools/lib/src/base/build.dart

```diff
@@ -0,0 +1,44 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert' show JSON;
+
+import 'package:crypto/crypto.dart' show md5;
+
+import 'file_system.dart';
+
+/// A collection of checksums for a set of input files.
+///
+/// This class can be used during build actions to compute a checksum of the
+/// build action inputs, and if unchanged from the previous build, skip the
+/// build step. This assumes that build outputs are strictly a product of the
+/// input files.
+class Checksum {
+  Checksum.fromFiles(Set<String> inputPaths) : _checksums = <String, int>{} {
+    final List<File> files = inputPaths.map(fs.file).toList(growable: false);
+    final List<File> missingInputs = files.where((File file) => !file.existsSync()).toList();
+    if (missingInputs.isNotEmpty)
+      throw new ArgumentError('Missing input files:\n' + missingInputs.join('\n'));
```

Comment:
Can you think of a better exception here? `ArgumentError` feels pretty borderline.